flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java

```diff
@@ -119,10 +153,50 @@ public void shutdown() {
 		}
 	}
 
+	/**
+	 * Registers and pins a cached ResultPartition that holds the data for an IntermediateResultPartition.
+	 * @param partitionID The IntermediateResultPartitionID to find a corresponding ResultPartition for.
+	 * @param numConsumers The number of consumers that want to access the ResultPartition
+	 * @return true if the registering/pinning succeeded, false otherwise.
+	 */
+	public boolean pinCachedResultPartition(IntermediateResultPartitionID partitionID, int numConsumers) {
+		synchronized (cachedResultPartitions) {
+			ResultPartition resultPartition = cachedResultPartitions.get(partitionID);
+			if (resultPartition != null) {
+				try {
+					// update its least recently used value
+					updateIntermediateResultPartitionCache(resultPartition);
+
+					synchronized (registeredPartitions) {
+						if (!registeredPartitions.containsValue(resultPartition)) {
+							LOG.debug("Registered previously cached ResultPartition {}.", resultPartition);
+							registerResultPartition(resultPartition);
```

Comment:
I think there is a race between pinning and releasing. It could happen that a pinned result is removed by a concurrent release (in the onConsumed callback).